torch/csrc/distributed/c10d/Ops.cpp

```diff
@@ -310,6 +268,12 @@ c10::intrusive_ptr<Work> broadcast(
     const c10::intrusive_ptr<ProcessGroup>& process_group,
     at::TensorList tensors,
     const BroadcastOptions& opts) {
+  // TODO: handles the case of using a PythonProcessGroup
+  if (!process_group->hasBackends()) {
+    auto tensor_vec = tensors.vec();
+    return process_group->broadcast(tensor_vec, opts);
+  }
+
```

Comment:
**TLDR: without this change any pure PythonProcessGroup extensions (e.g. XLA) would fail DDP tests.**

The reason is that `Reducer.cpp` calls `Ops.cpp` directly. Reducer uses `allreduce`, `broadcast`, and `allgather` which is why only those collectives have this check.

For the PythonProcessGroup extension case, we are not registering backends https://github.com/pytorch/pytorch/pull/88330/files#diff-61109d1cb2a0bd13fc51d678a82666295289da1ec0a1a694e73d9e8c28f51bdcR1040-R1045 and defaulting to the implementation provided by the user. 

Here is the workflow of how the PythonProcessGroup:

init_process_group(backend="CustomPythonPG") sets CustomPythonPG as default process group -> `DDP()` uses the default process group -> `Reducer` is called by DDP -> Reducer calls  a collective, such as `ops.broadcast` -> this broadcast calls from C++ via PYBIND11_OVERRIDE (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/PyProcessGroup.hpp#L90-L95) into the Python definition of broadcast in CustomPythonPG

Heres why it breaks with the dispatcher implementation:

init_process_group(backend="CustomPythonPG") sets CustomPythonPG as default process group -> `DDP()` uses the default process group -> `Reducer` is called by DDP -> Reducer calls  a collective, such as `ops.broadcast` -> 

Everything is same up to this point, however now `ops.broadcast` will to dispatch into a backend implementation such as `broadcast_cuda_` or `broadcast_cpu_` which errors out since there is no backend defined.

Reply:
Update this comment to explain this is for the PythonProcessGroup reducer test case and link to the issue